### PR TITLE
Make FakeValuesService expression caches static to share across Faker instances

### PR DESCRIPTION
**Pull request by mferretti**
## Problem

Every `new Faker()` starts with empty expression-resolution and regex-compilation caches. In workloads that create many short-lived Fakers — property-based testing frameworks (jqwik, QuickCheck), seeded-per-record data generation — the warm-up cost is paid repeatedly, even when Fakers share the same locale.

## Why "just make the map static" needed structural work

The existing `REGEXP2SUPPLIER_MAP` used `RegExpContext(String exp, BaseFaker root, FakerContext context)` as its cache key. Making it static would re-introduce the memory leak fixed in #1263 / PR #1271: the `root` field holds a strong reference to the `Faker` instance, so every `new Faker()` adds a unique entry the GC can never collect.

Fixing this requires separating *what method to call* (shareable, context-free) from *which provider instance to call it on* (per-Faker, bound at first use).

## What changed

**Two-level expression cache:**
- **L1 `RECIPE_MAP` (static)** — keyed by `CacheKey(String exp, SingletonLocale locale)`. No per-Faker references; safe to share globally. Stores context-free "recipe" resolvers.
- **L2 `instanceMap` (per-instance)** — stores "materialized" resolvers pre-bound to this Faker's concrete provider instances. Subsequent calls on the same Faker skip `getProvider()` entirely.

`ValueResolver` gains `materialize(ProviderRegistration root)` and `cacheable()` to support the two-level contract. New resolver types (`ProviderMethodResolver`, `ChainedCoercedResolver`, `InstanceMethodResolver`, etc.) are context-free at L1 and pre-bound at L2.

**`expression2generex` made static** — the `RgxGen` compiled-regex cache was per-instance. This is the primary performance win: every new Faker recompiled the same patterns; now compilation happens once globally.

## No public API changes

`FakeValuesService` remains internal. `Faker` constructors are unchanged. No new public methods.

## Tests

`SharedFakeValuesServiceTest` covers:
- 16 threads × 10k iterations of concurrent Faker creation — no errors, no races
- 4 locales × 4 threads concurrent — no errors
- Same seed → same output regardless of static cache state

### CODE REVIEW

File: src/main/java/net/datafaker/service/FakeValuesService.java [hunk 1]
```diff
@@ -83,7 +84,18 @@ public class FakeValuesService {
...
+    private static final ConcurrentHashMap<Locale, FakeValuesService> SHARED_INSTANCES = new ConcurrentHashMap<>();
+
     private final Map<RegExpContext, ValueResolver> REGEXP2SUPPLIER_MAP = new CopyOnWriteMap<>(HashMap::new);
+
+    /**
+     * Returns a lazily-initialized per-locale singleton. Safe to share across threads:
+     * all mutable instance state uses idempotent copy-on-write caches.
+     */
+    public static FakeValuesService getShared(Locale locale) {
+        return SHARED_INSTANCES.computeIfAbsent(locale, l -> new FakeValuesService());
```
[the remark this answers is not in the document]
Comment (mferretti) [reply src/main/java/net/datafaker/service/FakeValuesService.java]: The locale isn't passed to the constructor — it doesn't need to be. Looking at updateFakeValuesInterfaceMap (line 85–88), locale-specific YAML is loaded lazily into fakeValuesInterfaceMap after construction, triggered by the BaseFaker constructor. The Locale key in SHARED_INSTANCES does exactly what it needs to: it ensures all threads using Locale.ENGLISH share one FakeValuesService instance which then lazily accumulates the English data once. Locale.FRENCH gets a separate one. The whole point is that the n-thread scenario from the issue — same locale, different Random — gets one service loaded once.

File: src/main/java/net/datafaker/service/FakeValuesService.java [hunk 2]
```diff
@@ -83,7 +84,18 @@ public class FakeValuesService {
...
+    }
```
[the remark this answers is not in the document]
Comment (mferretti) [reply src/main/java/net/datafaker/service/FakeValuesService.java]: Correct. Adding a guard against this scenario on shared instances and updating the javadoc accordingly

**Comment by asolntsev** [comment]
@mferretti Please don't take my words as critics, I am just researching the problem.

I feel that this solution heavily depends on the internal structure of DF. Ideally, the end-user should use only `new Faker()`, and shouldn't even know any its internals like `FakeValuesService` etc.

The bottleneck is redundant YAML loading, right? **If we want to avoid redundant YAML loading, then shouldn't we just extract the YML loading to some smaller class, and cache its results?**

**Comment by mferretti** [comment]
@asolntsev , none taken :) 
It  was food for thought and made me go back at the code.

YAML loading is not the bottleneck — it's already cached globally in `FakeValues.of()`. The real issue is regex compilation: `FakeValuesService` has a `REGEXP2SUPPLIER_MAP` that is a per-instance field, so every `new Faker()` starts with an empty regex compilation cache.

You're right that exposing `FakeValuesService` is leaky. The cleaner fix is simply making `REGEXP2SUPPLIER_MAP` static (like `EXPRESSION_2_SPLITTED` already is) — users just call `new Faker()` and get shared regex caching for free, no API changes needed.

Would that make sense? If so, I can revise the PR along those lines :-)

**Comment by asolntsev** [comment]
@mferretti Yes, it seems absolutely reasonable to make `REGEXP2SUPPLIER_MAP` static. Seems it should solve the initial problem?

UPD I realized that it was previously static, and I made in non-static in commit https://github.com/datafaker-net/datafaker/pull/1271/changes/751ab16d25608c5d49aae1844359ac9b1c90b39e    :) That's an irony! :) 

The reason was [memory leak](https://github.com/datafaker-net/datafaker/issues/1263). Static map `REGEXP2SUPPLIER_MAP` was growing endlessly, collecting indefinite number of unique regexps. 

Maybe we should investigate https://github.com/datafaker-net/datafaker/issues/1263 once again and find a better solution for it. Or just use some Cache-like map for `REGEXP2SUPPLIER_MAP` (which removes unused records after some time).

**Comment by mferretti** [comment]
@asolntsev thanks for pointing this out. I'll have a look at the bug that made you opt for non static and try to put everything together. it'll have to wait a couple of days as i am heading out and will be back tomorrow late evening/night.

**Comment by kingthorin** [comment]
No rush, enjoy your weekend!!

**Comment by mferretti** [comment]
Hello @asolntsev and @kingthorin  
here's my understanding/findings (I am not going to question the why of the changes or the architecture/development, just exposing what I saw and understood)
- Every new Faker() starts with an empty REGEXP2SUPPLIER_MAP — the cache that records which method resolves each YAML expression (e.g. #{Name.firstName} → Name.firstName()). In high-volume scenarios where many Faker instances are created, method-resolution work is repeated from scratch for every instance, causing slowdown.
-  REGEXP2SUPPLIER_MAP was previously static but was made per-instance in PR #1271 to fix a memory leak (HashMap grew without bound as every Faker produced unique cache keys and entries never evicted).

What I propose, but i'd need some validation from you guys, is :
1. expression2generex becomes static (does not help yaml resolution but helps regexify)
2. add a static map and implement a "context free" ValueResolver

On point 2, `resolve` is private so there shouldn't be any breaking changes
On point 1: the main caveat here is the assumption that the provider is registred with the same name across all the Faker's root; probably an edge case but custom providers could be registering via non standard names thus i'd add a test case.

 
Obviously, I can recycle this pr or close this one and open a new one. 
 
 Looking forward for your input !

**Comment by asolntsev** [comment]
Hi @mferretti !
Yes, it seems like a totally valid plan:
* expression2generex becomes static (does not help yaml resolution but helps regexify)
* add a static map and implement a "context free" ValueResolver

**Comment by mferretti** [comment]
Hi. 
sorry for the delay... life happens. 
I had to revert what i done in the first 2 commits as we are changing the view radically on what needs to be done. 

The original PR added a public `FakeValuesService.getShared(Locale)` API and `Faker.withSharedService()` to let callers share a single `FakeValuesService` across multiple Faker instances, reducing the cost of repeated cache warm-up. @asolntsev correctly pointed out that exposing `FakeValuesService` publicly is too leaky. His suggestion: make the internal map static instead.

This revised PR implements exactly that — but with the additional structural work required to do it correctly (naively making the map static brings back the memory leak from issue #1263 / PR #1271).
The fix requires two steps: 
1. Change the key to `CacheKey(String exp, SingletonLocale locale)` — no per-Faker references.
2. Separate which method to call (shareable) from which provider instance to call it on  (per-Faker). The `ValueResolver` implementations previously closed over per-Faker provider instances, so they could not be shared across Fakers.

Also, I added an L2 cache:

**L1 — static `RECIPE_MAP: Map<CacheKey, ValueResolver>`**
Stores context-free "recipes": which provider class and which method to invoke. Shared across all Fakers with the same locale. No per-Faker references → no memory leak.

**L2 — per-instance `instanceMap: Map<String, ValueResolver>`**
Stores "materialized" resolvers: the recipe pre-bound to this Faker's concrete provider object. Fast repeated calls on the same Faker skip the `getProvider()` reflective call entirely.

Resolution order: L2 hit → L1 hit (materialize + store L2) → full discovery (store L1 + materialize L2).

The benchmarks tests i did locally: 
| Scenario | `main` | This PR | Delta |
|---|---|---|---|
| 10k new Fakers × `vehicle().vin()` (heavy regex) | 75 ms | 7 ms | **10× faster** |
| 10k new Fakers × `finance().bic()` (medium regex) | 29 ms | 8 ms | **3.6× faster** |
| Single Faker × 10k `vehicle().vin()` | 6 ms | 5 ms | unchanged |
| 1000 new Fakers × 3 calls (name/address/internet) | 10,275 µs | 7,940 µs | **23% faster** |
| 10k single-shot Fakers × 1 call | 25,222 µs | 19,112 µs | **24% faster** |
| Single Faker × 1000 calls | 5,024 µs | 4,278 µs | unchanged |

In all honesty, I would suggest that  you run your tests  too as the performance gain, on my side, is clear with complex regex and a property based test scenario, but non existent with simple regex; plus  I have added the L2 cache complexity.

File: src/main/java/net/datafaker/service/FakeValuesService.java [hunk 3]
```diff
@@ -681,12 +699,12 @@ private String[] splitExpressions(String expression, int length) {
         });
     }
 
-    private Object resExp(String directive, String[] args, Object current, ProviderRegistration root, FakerContext context, RegExpContext regExpContext) {
+    private Object resExp(String directive, String[] args, Object current, ProviderRegistration root, FakerContext context, CacheKey cacheKey) {
```
Comment (kingthorin) [inline src/main/java/net/datafaker/service/FakeValuesService.java]: I know this was existing, I don't expect it to be changed here, it's just caught my eye 👁️ 

Should this actually be `regExp?

Comment (mferretti) [reply src/main/java/net/datafaker/service/FakeValuesService.java]: It's my fixation: `resExp` is short for "resolve expression" (not regex), but the old `RegExpContext regExpContext` parameter made it look regex-related. Here I am removing  `RegExpContext` entirely and replacing it with `CacheKey`, so the confusing parameter name disappears. Agreed it's out of scope to rename the method itself here, but happy to do it in a follow-up

Comment (kingthorin) [reply src/main/java/net/datafaker/service/FakeValuesService.java]: Let's actually use `resolveExpression` then. It's a private method, it won't be breaking and it will be clear for the future 😉

File: src/main/java/net/datafaker/service/FakeValuesService.java [hunk 4]
```diff
@@ -165,23 +170,21 @@ public String fetchString(String key, FakerContext context) {
...
             this.simpleDirective = simpleDirective;
-            this.context = context;
         }
 
         @Override
-        public Object resolve() {
-            return safeFetch(simpleDirective, context, null);
+        public Object resolve(ProviderRegistration root, FakerContext context) {
+            return root.fakeValuesService().safeFetch(simpleDirective, context, null);
         }
```
[the remark this answers is not in the document]
Comment (mferretti) [reply src/main/java/net/datafaker/service/FakeValuesService.java:194]: Fixed: added  guard at the top of .

File: src/main/java/net/datafaker/service/FakeValuesService.java [hunk 5]
```diff
@@ -936,8 +967,8 @@ private MethodAndCoercedArgs accessor(Class<?> clazz, final String accessorName,
         });
 
         Collection<Method> methods = classMethodsMap.getOrDefault(name, emptyList());
-        if (methods.isEmpty()) {
-            LOG.fine(() -> "Didn't find accessor named %s on %s with args %s".formatted(accessorName, clazz.getSimpleName(), Arrays.toString(args)));
+        if (methods == null) {
+            LOG.fine(() -> "Didn't accessor named %s on %s with args %s (methods=%s)".formatted(accessorName, clazz.getSimpleName(), Arrays.toString(args), null));
             return null;
         }
```
[the remark this answers is not in the document]
Comment (mferretti) [reply src/main/java/net/datafaker/service/FakeValuesService.java:984]: Fixed: changed `if (methods == null)` → `if (methods.isEmpty())` and corrected both "Didn't accessor" log messages to "Didn't find accessor".

File: src/test/java/net/datafaker/SharedFakeValuesServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class SharedFakeValuesServiceTest {
+
+    @Test
+    void concurrentFakersProduceNoErrors() throws Exception {
+        int threads = 16;
+        int iterations = 10_000;
```
[the remark this answers is not in the document]
Comment (mferretti) [reply src/test/java/net/datafaker/SharedFakeValuesServiceTest.java:25]: The PR title and description have been rewritten. The new description describes exactly what this PR does: makes `expression2generex` and the resolver caches static. No public API changes, no `getShared`/`withSharedService`.

File: src/main/java/net/datafaker/service/FakeValuesService.java [hunk 7]
```diff
@@ -733,19 +753,19 @@ private Object resolveExpression(String directive, String[] args, Object current
                 if (current instanceof AbstractProvider) {
                     final Method method = BaseFaker.getMethod((AbstractProvider<?>) current, directive);
                     if (method != null) {
-                        res.add(new MethodResolver(method, current, args));
+                        res.add(new ProviderMethodResolver(current.getClass().getSimpleName(), method, args));
                         return res;
                     }
```
[the remark this answers is not in the document]
Comment (mferretti) [reply src/main/java/net/datafaker/service/FakeValuesService.java:769]: Fixed: added `if (root == null) return this` in `ProviderMethodResolver.materialize()` and `if (root == null) return null` in its `resolve()`. Also guarded the outer materialize calls in `resolveExpression` so we skip L2 caching when root is null.

File: src/main/java/net/datafaker/service/FakeValuesService.java (hunk 3 above)
Comment (mferretti) [reply src/main/java/net/datafaker/service/FakeValuesService.java]: it's done :) ... together with a bunch of fixes and some improvements driven by the copilot checks

**Comment by kingthorin** [comment]
Seems reasonable to me, but someone that knows the code better should review or merge.

**Comment by mferretti** [comment]
@asolntsev , @kingthorin  good morning.
It's awfully hot in here so i woke up early to take advantange of the "cool 25C" early morning and took a look at the conflict.
The base issue derived from #1849 which reworked the same single map cache this PR replaces with the L1/L2 design. I kept the L1/L2 design and worked to keep compatibility with upstream and discovered that this PR actually had a gap as the L2 cache was keyed by expression only so a `FakerValueService` shared across `Faker`s could hand a resolver bound to one root to another; L2 entries now carry `(root, context)` and are reused only on identity match.
I added a test `SharedFakeValuesServiceTest#sharedServiceAcrossRootsKeepsPerRootDeterminism` that fails if the the guard introduced in #1849 is missing, namely checking if root B observes root A random stream.
I reran my performance tests taking into account a case where you have multiple locales and one with a single locale: 

| Workload | regex | this PR | main | speedup |
|---|---|---|---|---|
| `vehicle().vin()` | complex | 6.6 | 65.1 | ~10× |
| `finance().bic()` | medium | 8.0 | 29.2 | ~3.6× |
| `regexify("[a-z]{5}")` | simple | 2.6 | 9.1 | ~3.5× |
| single Faker × 10k `vin()` | steady | 7.2 | 8.3 | ~parity |

Times = median ms, 10k fresh Fakers same locale (last row = one Faker × 10k). Identical outputs both sides.

test for single locale: 
```java
import net.datafaker.Faker;
import java.util.Arrays;
import java.util.Locale;
import java.util.Random;

// Usage: java -cp <datafaker+deps> Bench <vin|bic|simple> <new|single> [n]
public class Bench {
    static long work(Faker f, String w) {
        return switch (w) {
            case "vin" -> f.vehicle().vin().length();
            case "bic" -> f.finance().bic().length();
            default    -> f.regexify("[a-z]{5}").length();
        };
    }
    static long modeNew(String w, int n) {           // n fresh Fakers, same locale
        long s = 0;
        for (int i = 0; i < n; i++) s += work(new Faker(Locale.ENGLISH, new Random(i)), w);
        return s;
    }
    static long modeSingle(String w, int n) {        // one Faker, steady state
        long s = 0; Faker f = new Faker(Locale.ENGLISH, new Random(1));
        for (int i = 0; i < n; i++) s += work(f, w);
        return s;
    }
    public static void main(String[] args) {
        String w = args.length > 0 ? args[0] : "vin";
        String mode = args.length > 1 ? args[1] : "new";
        int n = args.length > 2 ? Integer.parseInt(args[2]) : 10000;
        java.util.function.IntToLongFunction run =
            "single".equals(mode) ? k -> modeSingle(w, k) : k -> modeNew(w, k);
        long guard = 0;
        for (int i = 0; i < 5; i++) guard += run.applyAsLong(n);   // warmup
        double[] ms = new double[9];
        for (int t = 0; t < 9; t++) {
            long start = System.nanoTime();
            guard += run.applyAsLong(n);
            ms[t] = (System.nanoTime() - start) / 1_000_000.0;
        }
        Arrays.sort(ms);
        System.out.printf("%-7s %-7s n=%d  median=%6.1f ms  min=%6.1f ms  [guard=%d]%n",
            w, mode, n, ms[4], ms[0], guard);
    }
}
```
multiple locales: 

**Fresh Fakers — `new Faker(locale, …)` per call** (issue #1814: many short-lived Fakers)

| Workload | regex | this PR | main | speedup |
|---|---|---|---|---|
| `vehicle().vin()` | complex | 9.4 | 68.4 | ~7.3× |
| `finance().bic()` | medium | 12.8 | 32.2 | ~2.5× |
| `regexify("[a-z]{5}")` | simple | 5.6 | 12.5 | ~2.2× |

**Single long-lived Faker — locale switched per call via `doWith`** (steady state)

| Workload | regex | this PR | main | delta |
|---|---|---|---|---|
| `vehicle().vin()` | complex | 10.6 | 11.0 | ~parity |
| `finance().bic()` | medium | 7.6 | 12.2 | ~1.6x * |
| `regexify("[a-z]{5}")` | simple | 4.2 | 4.4 | ~parity |



Median ms, one Faker × 10k, locale rotated over 8 locales per call. Identical outputs both sides.
`*` bic() rebuilds its regex per country code → this PR's static compile-cache wins even on a single Faker.


test: 
```java
import net.datafaker.Faker;
import java.util.Arrays;
import java.util.Locale;
import java.util.Random;

// Usage: java -cp <datafaker+deps> Bench <vin|bic|simple> <new|single|multi|singlemulti> [n]
//   new          : n fresh Fakers, single locale          (warm-up amortization)
//   single       : one Faker, n iterations, single locale (steady state)
//   multi        : n fresh Fakers, locale rotated          (multi-locale warm-up)
//   singlemulti  : one Faker, locale switched per call     (multi-locale steady state)
public class Bench {
    private static final Locale[] LOCALES = {
        Locale.ENGLISH, Locale.FRENCH, Locale.GERMAN, Locale.ITALIAN,
        Locale.forLanguageTag("es"), Locale.forLanguageTag("pt"),
        Locale.forLanguageTag("nl"), Locale.forLanguageTag("pl")
    };

    static long work(Faker f, String w) {
        return switch (w) {
            case "vin" -> f.vehicle().vin().length();
            case "bic" -> f.finance().bic().length();
            default    -> f.regexify("[a-z]{5}").length();
        };
    }

    static long modeNew(String w, int n) {
        long s = 0;
        for (int i = 0; i < n; i++) s += work(new Faker(Locale.ENGLISH, new Random(i)), w);
        return s;
    }
    static long modeSingle(String w, int n) {
        long s = 0; Faker f = new Faker(Locale.ENGLISH, new Random(1));
        for (int i = 0; i < n; i++) s += work(f, w);
        return s;
    }
    static long modeMulti(String w, int n) {
        long s = 0;
        for (int i = 0; i < n; i++) s += work(new Faker(LOCALES[i % LOCALES.length], new Random(i)), w);
        return s;
    }
    static long modeSingleMulti(String w, int n) {
        long s = 0; Faker f = new Faker(Locale.ENGLISH, new Random(1));
        for (int i = 0; i < n; i++) {
            final Locale loc = LOCALES[i % LOCALES.length];
            s += f.doWith(() -> work(f, w), loc);
        }
        return s;
    }

    public static void main(String[] args) {
        String w = args.length > 0 ? args[0] : "vin";
        String mode = args.length > 1 ? args[1] : "new";
        int n = args.length > 2 ? Integer.parseInt(args[2]) : 10000;
        java.util.function.IntToLongFunction run = switch (mode) {
            case "single"      -> k -> modeSingle(w, k);
            case "multi"       -> k -> modeMulti(w, k);
            case "singlemulti" -> k -> modeSingleMulti(w, k);
            default            -> k -> modeNew(w, k);
        };
        long guard = 0;
        for (int i = 0; i < 5; i++) guard += run.applyAsLong(n);   // warmup
        double[] ms = new double[9];
        for (int t = 0; t < 9; t++) {
            long start = System.nanoTime();
            guard += run.applyAsLong(n);
            ms[t] = (System.nanoTime() - start) / 1_000_000.0;
        }
        Arrays.sort(ms);
        System.out.printf("%-7s %-12s n=%d  median=%6.1f ms  min=%6.1f ms  [guard=%d]%n",
            w, mode, n, ms[4], ms[0], guard);
    }
}
```

In the multiple locales scenario the gain is smaller but the winning key is the fact the the regular expression is compiled only once so, in a scenario with multiple Faker instances and locale, it's practically a constant; what actually changes in the multiple scenario case is the fact that YAML interface map AND L1 method resolution cache are locale aware so there we lose something.

**Comment by kingthorin** [comment]
I appreciate you staying on this, but I still think this needs more history/insight than I have for this code base.

**Comment by bodiam** [comment]
Same here. @snuyanzin , could you help out?

**Comment by snuyanzin** [comment]
it seems we are in a lack of tests for custom fakers
I added some of them and if you try this PR against these tests there will be failures https://github.com/datafaker-net/datafaker/pull/1859

Locally I have more failures, however need to prepare them as tests first

especially for cases where different faker might be registered under same simple name (different globals)

**Comment by mferretti** [comment]
Ok,
@asolntsev , @bodiam , @kingthorin  I’ll wait for the next stream of tests to be pushed upstream and go from there if you’re interested in the feature

**Comment by snuyanzin** [comment]
I played with code measured perf from comments
1. It does not show the mentioned improvements in my setup
2. error rate is too high

can you please share JMH benchmarks to be more precise?

**Comment by mferretti** [comment]
hi @snuyanzin 
sorry it took so long, your message caught me during a weekend trip :)
This post looks long but it's actually the longest task on my test  :rofl: :rofl: :rofl: as I was able to put together a throwaway test based on the code I used to produce the tables above. 
Here's the JMH visualizer [link](https://jmh.morethan.io/?gist=1ad6729a3779f7b8844d793edd593e0b) if you want to look at it before going down this detailed post and the [raw jmh data](https://gist.github.com/mferretti/1ad6729a3779f7b8844d793edd593e0b)

The setup: 
- JMH 1.37
- Average Time as µs/op
- 2 forks x (3 warmup + 5 measures x 1s)
- speedup = main / this PR.


The code: 
I am calling public API only, so the bytecode is identical both on main and this PR; the only thing that changes is DataFaker on the run classpath.

The runs:  since JMH can only load one version of `net.datafaker` 
1. I have compiled the test in order to keep the same bytecode
2. Ran the same out/ twice, swapping only the datafaker classes on the classpath:

`java -cp out:<PR classes>:<deps>:<jmh>  org.openjdk.jmh.Main -rff ours.json`
here  PR classes = /home/marco/development/datafaker/target/classes (HEAD 5cfffe78)

`java -cp out:<main classes>:<deps>:<jmh>  org.openjdk.jmh.Main -rff main.json`
here main classes = /tmp/df-upstream/target/classes (HEAD 7492624f)



```java
package bench;

import net.datafaker.Faker;
import org.openjdk.jmh.annotations.*;

import java.util.Locale;
import java.util.Random;
import java.util.concurrent.TimeUnit;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
@Warmup(iterations = 3, time = 1)
@Measurement(iterations = 5, time = 1)
@Fork(2)
@State(Scope.Thread)
public class Bench {

    @Param({"vin", "bic", "simple"})
    public String workload;

    private static final Locale[] LOCALES = {
        Locale.ENGLISH, Locale.FRENCH, Locale.GERMAN, Locale.ITALIAN,
        Locale.forLanguageTag("es"), Locale.forLanguageTag("pt"),
        Locale.forLanguageTag("nl"), Locale.forLanguageTag("pl")
    };

    private int seed;
    private Faker single;

    @Setup
    public void setup() {
        single = new Faker(Locale.ENGLISH, new Random(1));
    }

    static long work(Faker f, String w) {
        return switch (w) {
            case "vin" -> f.vehicle().vin().length();
            case "bic" -> f.finance().bic().length();
            default    -> f.regexify("[a-z]{5}").length();
        };
    }

    @Benchmark
    public long freshFaker() {                       // issue #1814: short-lived Faker each call
        return work(new Faker(Locale.ENGLISH, new Random(seed++)), workload);
    }

    @Benchmark
    public long singleFaker() {                      // steady state
        return work(single, workload);
    }

    @Benchmark
    public long freshFakerMultiLocale() {            // short-lived Faker, locale rotated
        Locale loc = LOCALES[(seed++ & 7)];
        return work(new Faker(loc, new Random(seed)), workload);
    }

    @Benchmark
    public long singleFakerMultiLocale() {           // one Faker, locale switched via doWith
        Locale loc = LOCALES[(seed++ & 7)];
        return single.doWith(() -> work(single, workload), loc);
    }
}
```

The results I got on my laptop for the scenarios: 
`▎ Hardware: AMD Ryzen 5 PRO 4650U (6 cores / 12 threads), 32 GB RAM, Ubuntu 24.04 (kernel 6.17), Amazon Corretto JDK 21.0.9 (OpenJDK 64-Bit Server VM).`


`Fresh short-lived Faker each call, ENGLISH (issue #1814 — the target scenario)`
| Workload | regex | this PR | main | speedup |
|---|---|---|---|---|
| `vehicle().vin()` | complex | 0.729 ± 0.015 | 7.059 ± 0.324 | **~9.7×** |
| `finance().bic()` | medium | 0.803 ± 0.024 | 2.970 ± 0.035 | **~3.7×** |
| `regexify("[a-z]{5}")` | simple | 0.282 ± 0.011 | 0.934 ± 0.010 | **~3.3×** |

`Fresh Faker each call, 8 locales rotated`
| Workload | regex | this PR | main | speedup |
|---|---|---|---|---|
| `vehicle().vin()` | complex | 0.960 ± 0.042 | 7.135 ± 0.315 | **~7.4×** |
| `finance().bic()` | medium | 1.211 ± 0.061 | 3.529 ± 0.044 | **~2.9×** |
| `regexify("[a-z]{5}")` | simple | 0.486 ± 0.002 | 1.143 ± 0.051 | **~2.4×** |

`One long-lived Faker, ENGLISH (steady state)`
| Workload | regex | this PR | main | delta |
|---|---|---|---|---|
| `vehicle().vin()` | complex | 0.531 ± 0.011 | 0.540 ± 0.009 | parity |
| `finance().bic()` | medium | 0.440 ± 0.008 | 0.437 ± 0.004 | parity |
| `regexify("[a-z]{5}")` | simple | 0.095 ± 0.003 | 0.097 ± 0.002 | parity |

`One Faker, locale switched per call via `doWith`
| Workload | regex | this PR | main | delta |
|---|---|---|---|---|
| `vehicle().vin()` | complex | 0.643 ± 0.013 | 0.662 ± 0.008 | parity |
| `finance().bic()` | medium | 0.714 ± 0.022 | 0.687 ± 0.025 | parity |
| `regexify("[a-z]{5}")` | simple | 0.172 ± 0.007 | 0.171 ± 0.006 | parity |


As far as I can derive from this test: 
- the win on the short-lived faker  stays on the complex  scenario (bigest win) through the simple scenario.
- single long-lived faker JMH/ shows parity on every type of workload as expected: the above win kicks in during multiple regex compilations

_Correction on my previous report_ : 
I had reported `finance().bic()` on a single Faker with rotating locales as ~1.6×. JMH puts that at parity (0.714 vs 0.687) . Evidently the 1.6× was measurement noise  which I think it's the imprecision you flagged. The static regex-compile cache only helps when the compile cost would otherwise be paid again on a fresh instance.

**Comment by mferretti** [comment]
Hello @asolntsev @kingthorin  and @bodiam . 
I am not sure if I should look into the conflicts: does this pr have any chance to make it through ? Are there issues I should be aware of before investing more into this ? 
TIA

**Comment by kingthorin** [comment]
I'm sorry but I still don't know the code base and performance impacts well enough to say one way or another.

If I recall correctly I did previously note that the jmh error numbers got worse with this change but wasn't able to see why or dig in to it.

I'm all for improving performance but I don't think I can make the call here.

**Comment by snuyanzin** [comment]
based on taking a look there are still several sync issues in current repo which in my opinion should be fixed first

however unfortunately don't have time for this

**Comment by mferretti** [comment]
@kingthorin I am not 100% sure what you're talking about: is there some JMH packages for the project I can test with?
@snuyanzin I don't understand what you mean by "several sync issues in current repo". Anything i can help with ?

Said that, I can fix this PR and align it at current master, just need to find a few hours in which i can work on it
